vsphere/check.py

```diff
@@ -924,6 +957,297 @@ def collect_metrics(self, instance):
+
+        def build_id(vsphere_url, object_type, object_name):
+            return "urn://vsphere/{0}/{1}/{2}".format(vsphere_url, object_type, object_name)
+
+        def build_type(object_type):
+            return {"name": object_type}
+
+        topology_items = self.get_topologyitems_sync(instance)
+        vsphere_url = instance.get("host")
+        instance_key = {"type": self.INSTANCE_TYPE, "url": vsphere_url}
+
+        self.start_snapshot(instance_key)
+
+        for vm in topology_items["vms"]:
+            self.component(
+                instance_key,
+                build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.VM, vm["hostname"]),
+                build_type(VSPHERE_COMPONENT_TYPE.VM) ,
+                vm["topo_tags"]
+            )
+
+        for host in topology_items["hosts"]:
+            self.component(
+                instance_key,
+                build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.HOST, host["hostname"]),
+                build_type(VSPHERE_COMPONENT_TYPE.HOST),
+                host["topo_tags"]
+            )
+            for vm_id in host["topo_tags"]["vms"]:
+                self.relation(
+                    instance_key,
+                    build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.VM, vm_id),
+                    build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.HOST, host["hostname"]),
+                    build_type(VSPHERE_RELATION_TYPE.VM_HOST)
+                )
+
+
+        for cluster in topology_items["clustercomputeresource"]:
+            self.component(
+                instance_key,
+                build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.CLUSTERCOMPUTERESOURCE, cluster["topo_tags"]["name"]),
+                build_type(VSPHERE_COMPONENT_TYPE.CLUSTERCOMPUTERESOURCE),
+                cluster["topo_tags"]
+            )
+
+        for cluster in topology_items["computeresource"]:
+            self.component(
+                instance_key,
+                build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.COMPUTERESOURCE, cluster["topo_tags"]["name"]),
+                build_type(VSPHERE_COMPONENT_TYPE.COMPUTERESOURCE),
+                cluster["topo_tags"]
+            )
+
+            for host_id in cluster["topo_tags"]["hosts"]:
+                self.relation(
+                    instance_key,
+                    build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.HOST, host_id),
+                    build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.COMPUTERESOURCE, cluster["topo_tags"]["name"]),
+                    build_type(VSPHERE_RELATION_TYPE.HOST_COMPUTERESOURCE),
+                    {}
+                )
+
+
+        for dc in topology_items["datacenters"]:
+            self.component(
+                instance_key,
+                build_id(vsphere_url, VSPHERE_COMPONENT_TYPE.DATACENTER, dc["topo_tags"]["name"]),
+                build_type(VSPHERE_COMPONENT_TYPE.DATACENTER),
+                dc["topo_tags"]
+            )
+            for ds_id in dc["topo_tags"]["datastores"]:
```

Comment:
variable `ds_id` is unused, does the iteration still make sense?

Reply:
fixed now.

Reply:
👍